### ISSUE TITLE
Bug: Defining functions with typescript complains of unused variable

### ISSUE THREAD

**Issue by justingolden21** [eslint/eslint#17039]
### Environment

Node version: 8.13.0
npm version: 8.19.3
Local ESLint version: 
```
		"@typescript-eslint/eslint-plugin": "^5.45.0",
		"@typescript-eslint/parser": "^5.45.0",
		"eslint": "^8.36.0",
		"eslint-config-prettier": "^8.8.0",
		"eslint-plugin-import": "^2.27.5",
		"eslint-plugin-prettier": "^4.2.1",
		"eslint-plugin-svelte3": "^4.0.0",
```
Global ESLint version: ?
Operating System: Windows 10

### What parser are you using?

Default (Espree)

### What did you do?

<details>
<summary>Configuration</summary>

```cjs
module.exports = {
	root: true,
	parser: '@typescript-eslint/parser',
	extends: ['eslint:recommended', 'plugin:@typescript-eslint/recommended', 'prettier'],
	plugins: ['prettier', 'svelte3', '@typescript-eslint'],
	ignorePatterns: ['*.cjs'],
	overrides: [
		{
			files: ['*.svelte'],
			processor: 'svelte3/svelte3'
		}
	],
	settings: {
		'svelte3/typescript': () => require('typescript')
	},
	parserOptions: {
		sourceType: 'module',
		ecmaVersion: 2020
	},
	env: {
		browser: true,
		es2017: true,
		node: true
	},
	globals: {
		gtag: true,
		__version__: true
	},
	rules: {
		'prettier/prettier': 'error',
		'quotes': ['warn', 'single', { avoidEscape: true }],
		'no-unused-vars': 'warn',
		'no-console': 'off',
		'no-tabs': 'off',
		'no-restricted-syntax': 'off',
		'func-names': 'off',
		'quote-props': ['error', 'consistent-as-needed'],
		'guard-for-in': 'off',
		'import/extensions': 'off',
		'new-cap': 'off',
		'import/no-extraneous-dependencies': 'off',
		'no-param-reassign': 'off',
		'radix': 'off',
		'@typescript-eslint/no-inferrable-types': 'off',

		// TODO: remove below
		'@typescript-eslint/no-explicit-any': 'off',
		'@typescript-eslint/ban-ts-comment': 'off'
	}
};

```
</details>

```ts
type Callback = (event: MouseEvent) => void;
const clickOutside = (node: Node, callback: Callback) => {
	const handleClick = (event: MouseEvent) => {
		if (node && !node.contains(event.target as Node) && !event.defaultPrevented) {
			/* dispatch event on click outside of node */
			callback(event);
		}
	};

	document.addEventListener('click', handleClick, true);

	return {
		destroy() {
			document.removeEventListener('click', handleClick, true);
		}
	};
};

export default clickOutside;
```

### What did you expect to happen?

No error

### What actually happened?

> 'event' is defined but never used.eslint[no-unused-vars](https://eslint.org/docs/rules/no-unused-vars)

### Participation

- [ ] I am willing to submit a pull request for this issue.

### Additional comments

![image](https://user-images.githubusercontent.com/30274440/228697767-15710e94-a2f3-4ddb-ba90-26a38313a154.png)

![image](https://user-images.githubusercontent.com/30274440/228697787-e9b16d1d-b78a-44c4-8257-555ece181d5b.png)

![image](https://user-images.githubusercontent.com/30274440/228697813-cddef0c8-816a-4d7c-bda0-54519a621e1a.png)

Even if I put it inline, it complains:

![image](https://user-images.githubusercontent.com/30274440/228697832-dfcab81d-ff48-4e3d-99a5-b968e30c2e1c.png)

One line reproducible example:

![image](https://user-images.githubusercontent.com/30274440/228697901-5690a15a-980a-4eea-bc67-90faf0a9b24d.png)

**Comment by aladdin-add** [comment]
the rule was not designed to work with ts,  you need to turn it off, and use [@typescript-eslint/no-unused-vars](https://typescript-eslint.io/rules/no-unused-vars) instead.
